### PR TITLE
fix m1 python 3.8 wasm issue

### PR DESCRIPTION
**Pull request by cqc-melf**

### CODE REVIEW

**Comment by cqc-melf** [comment]
Some tests are still failing because of: https://github.com/CQCL/tket/pull/571

**Comment by cqc-melf** [comment]
(And I have seen that python 3.10 has the same problem, so I am skipping that as well. :( )

**Comment by cqc-alec** [comment]
Following the discussion today I am inclined to abandon wasmer.

File: .github/workflows/build_and_test.yml [hunk 1]
```diff
@@ -254,7 +254,7 @@ jobs:
     - name: Install pybind11
       run: conan create --profile=tket recipes/pybind11
     - name: Build pytket (3.8)
-      if: github.event_name == 'push'
+      if: github.event_name == 'pull_request'
```
Comment (cqc-alec) [inline .github/workflows/build_and_test.yml:257]: Presumably this change (and below) should be reverted.

File: pytket/pytket/wasm/wasm.py [hunk 2]
```diff
@@ -16,24 +16,182 @@
...
+
+        # stores the names of the functions mapped
+        #  to the number of parameters and the number of return values
+        self._functions = dict()
+
+        # contains the list of functions that are not allowed
+        # to use in pytket (because of types that are not i32)
+        self._unsupported_function = []
+
+        # only run the wasmer import and so on if the check is requested to run
```
Comment (cqc-alec) [inline pytket/pytket/wasm/wasm.py:80]: Is this comment out of date? There is no wasmer now?

File: pytket/pytket/wasm/wasm.py [hunk 3]
```diff
@@ -16,24 +16,182 @@
...
+    }
+
+    def __init__(self, filepath: str, check_file: bool = True):
+        """construct a wasm file handler
+        :param filepath: path to the wasm file
+        :type filepath: str
+        :param check_file: If this parameter is set to True the wasm
+        file will be checked if the funtion signatures are supported.
+        This check is currently NOT supported on M1 mac in
+        python 3.8 and 3.10.
```
Comment (cqc-alec) [inline pytket/pytket/wasm/wasm.py:52]: Is this still true?

File: pytket/pytket/wasm/wasm.py [hunk 4]
```diff
@@ -16,24 +16,182 @@
...
+        """checks a given function name and signature if it is included
+        :param function_name: name of the function that is checked
+        :type function_name: str
+        :param number_of_parameters: number of i32 parameters of the function
+        :type number_of_parameters: int
+        :param number_of_returns: number of i32 return values of the function
+        :type number_of_returns: int
+        :return: true if the signature and the name of the function is correct"""
+        if not self._check_file:
+            return True
```
Comment (cqc-alec) [inline pytket/pytket/wasm/wasm.py:192]: Perhaps this should raise an exception rather than return True?

File: pytket/setup.py [hunk 5]
```diff
@@ -254,6 +254,7 @@ def finalize_options(self):
         "jinja2 ~= 3.0",
         "types-pkg_resources",
         "typing-extensions ~= 4.2",
+        "qwasm ~= 1.0",
```
Comment (cqc-alec) [inline pytket/setup.py:257]: Is qwasm open-source?

**Review by cqc-alec — CHANGES_REQUESTED** [review]

File: pytket/pytket/wasm/wasm.py (hunk 3 above)
Comment (cqc-melf) [reply pytket/pytket/wasm/wasm.py:52]: No, I will remove this.

File: pytket/pytket/wasm/wasm.py (hunk 4 above)
Comment (cqc-melf) [reply pytket/pytket/wasm/wasm.py:192]: The user can't choose if the check is executed. I was thinking about if this hole option to opt out of the check should be removed...

File: pytket/setup.py (hunk 5 above)
Comment (cqc-melf) [reply pytket/setup.py:257]: Yes:
https://github.com/CQCL/qwasm
https://pypi.org/project/qwasm/

File: .github/workflows/build_and_test.yml (hunk 1 above)
Comment (cqc-melf) [reply .github/workflows/build_and_test.yml:257]: Yes

File: pytket/pytket/wasm/wasm.py (hunk 4 above)
Comment (cqc-melf) [reply pytket/pytket/wasm/wasm.py:192]: I think we don't need it, it is only making thinks more complicated and in-transparent for the user.

File: pytket/setup.py (hunk 5 above)
Comment (cqc-alec) [reply pytket/setup.py:257]: Cool, would be good to add a link to the repo from the pypi page.

**Review by cqc-alec — COMMENTED** [review]
Happy to approve for merge once https://github.com/CQCL/tket/actions/runs/3188604820 has completed.

**Comment by cqc-melf** [comment]
I have added the link on https://pypi.org/project/qwasm/

**Review by cqc-alec — APPROVED** [review]

**Pull request merged**